source/jobs/JobsFeature.cpp

```diff
@@ -534,121 +528,137 @@ bool JobsFeature::isDuplicateNotification(JobExecutionData job)
-    {
-        for (Aws::Crt::JsonView view : jobDoc.GetArray(JOB_ATTR_ARGS))
-        {
-            args.push_back(view.AsString().c_str());
-            argsStringForLogging << view.AsString().c_str() << " ";
-        }
-    }
-    else
-    {
-        LOG_INFO(
-            TAG, "Did not find any arguments in the incoming job document. Value should be a JSON array of arguments");
-    }
-
-    if (operation.empty())
-    {
-        LOG_ERROR(TAG, "Unable to execute job, no operation provided!");
-        publishUpdateJobExecutionStatus(job, {Iotjobs::JobStatus::FAILED, "No operation provided", "", ""});
-        return;
-    }
-
-    ostringstream command;
+    ostringstream commandStream;
     bool operationOwnedByDeviceClient = false;
     if (DEFAULT_PATH_KEYWORD == path)
     {
         LOGM_DEBUG(TAG, "Using DC default command path {%s} for command execution", Sanitize(jobHandlerDir).c_str());
         operationOwnedByDeviceClient = true;
-        command << jobHandlerDir;
+        commandStream << jobHandlerDir;
     }
     else if (!path.empty())
     {
         LOGM_DEBUG(
             TAG, "Using path {%s} supplied by job document for command execution", Sanitize(path.c_str()).c_str());
-        command << path;
+        commandStream << path;
     }
     else
     {
         LOG_DEBUG(TAG, "Assuming executable is in PATH");
     }
-    command << operation;
+    commandStream << operation;
 
     if (operationOwnedByDeviceClient)
     {
-        const int actualPermissions = FileUtils::GetFilePermissions(command.str().c_str());
+        const int actualPermissions = FileUtils::GetFilePermissions(commandStream.str().c_str());
         if (Permissions::JOB_HANDLER != actualPermissions)
         {
             string message = FormatMessage(
                 "Unacceptable permissions found for job handler %s, permissions should be %d but found %d",
-                Sanitize(command.str()).c_str(),
+                Sanitize(commandStream.str()).c_str(),
                 Permissions::JOB_HANDLER,
                 actualPermissions);
             LOG_ERROR(TAG, message.c_str());
-            publishUpdateJobExecutionStatus(job, {Iotjobs::JobStatus::FAILED, message, "", ""});
-            return;
+            throw std::runtime_error(message);
         }
     }
 
-    int allowStdErr = 0;
-    if (jobDoc.KeyExists(JOB_ATTR_ALLOW_STDERR))
-    {
-        allowStdErr = jobDoc.GetInteger(JOB_ATTR_ALLOW_STDERR);
-    }
+    return commandStream.str();
+}
 
-    LOGM_DEBUG(
-        TAG, "About to execute: %s %s", Sanitize(command.str()).c_str(), Sanitize(argsStringForLogging.str()).c_str());
-    unique_ptr<JobEngine> engine(new JobEngine);
-    int executionStatus = engine->exec_cmd(command.str().c_str(), args);
+void JobsFeature::executeJob(JobExecutionData job)
+{
+    LOGM_INFO(TAG, "Executing job: %s", job.JobId->c_str());
+    publishUpdateJobExecutionStatus(job, {Iotjobs::JobStatus::IN_PROGRESS, "", "", ""});
```

Comment:
What if the this API call fails? shouldn't the body of the `runJob` lambda be part of the onComplete callback for `publishUpdateJobExecutionStatus`?

Reply:
In this instance, the publishUpdateJobExecutionStatus method performs an exponential backoff while attempting to update the status to inProgress. I didn't want to make execution of the job dependent on being able to set the job status to in progress, since it's not a hard requirement for us to update the execution status to in progress in order to eventually reach a terminal state for the job execution.

Reply:
I'm not clear on would constitute a hard requirement but in our docs for jobs for the device workflow, we require the device to update the job execution status before executing the job action

https://docs.aws.amazon.com/iot/latest/developerguide/jobs-devices.html#jobs-workflow-device-online